File: Firestore/Swift/Source/SwiftAPI/Stages.swift
/*
 * Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import Foundation

#if SWIFT_PACKAGE
  @_exported import FirebaseFirestoreInternalWrapper
#else
  @_exported import FirebaseFirestoreInternal
#endif // SWIFT_PACKAGE

@available(iOS 13, tvOS 13, macOS 10.15, macCatalyst 13, watchOS 7, *)
protocol Stage {
  var name: String { get }
  var bridge: StageBridge { get }
}

@available(iOS 13, tvOS 13, macOS 10.15, macCatalyst 13, watchOS 7, *)
class CollectionSource: Stage {
  let name: String = "collection"

  let bridge: StageBridge
  private var collection: CollectionReference
  private let db: Firestore

  init(collection: CollectionReference, db: Firestore) {
    self.collection = collection
    self.db = db
    bridge = CollectionSourceStageBridge(ref: collection, firestore: db)
  }
}

@available(iOS 13, tvOS 13, macOS 10.15, macCatalyst 13, watchOS 7, *)
class CollectionGroupSource: Stage {
  let name: String = "collectionId"

  let bridge: StageBridge
  private var collectionId: String

  init(collectionId: String) {
    self.collectionId = collectionId
    bridge = CollectionGroupSourceStageBridge(collectionId: collectionId)
  }
}

// Represents the entire database as a source.
@available(iOS 13, tvOS 13, macOS 10.15, macCatalyst 13, watchOS 7, *)
class DatabaseSource: Stage {
  let name: String = "database"
  let bridge: StageBridge

  init() {
    bridge = DatabaseSourceStageBridge()
  }
}

// Represents a list of document references as a source.
@available(iOS 13, tvOS 13, macOS 10.15, macCatalyst 13, watchOS 7, *)
class DocumentsSource: Stage {
  let name: String = "documents"
  let bridge: StageBridge
  private var docs: [DocumentReference]
  private let db: Firestore

  // Initialize with an array of String paths
  init(docs: [DocumentReference], db: Firestore) {
    self.docs = docs
    self.db = db
    bridge = DocumentsSourceStageBridge(documents: docs, firestore: db)
  }
}

// Represents an existing Query as a source.
@available(iOS 13, tvOS 13, macOS 10.15, macCatalyst 13, watchOS 7, *)
class QuerySource: Stage {
  let name: String = "query"
  let bridge: StageBridge
  private var query: Query

  init(query: Query) {
    self.query = query
    bridge = DatabaseSourceStageBridge()
    // TODO: bridge = QuerySourceStageBridge(query: query.query)
  }
}

// Represents an existing AggregateQuery as a source.
@available(iOS 13, tvOS 13, macOS 10.15, macCatalyst 13, watchOS 7, *)
class AggregateQuerySource: Stage {
  let name: String = "aggregateQuery"
  let bridge: StageBridge
  private var aggregateQuery: AggregateQuery

  init(aggregateQuery: AggregateQuery) {
    self.aggregateQuery = aggregateQuery
    bridge = DatabaseSourceStageBridge()
    // TODO: bridge = AggregateQuerySourceStageBridge(aggregateQuery: aggregateQuery.query)
  }
}

@available(iOS 13, tvOS 13, macOS 10.15, macCatalyst 13, watchOS 7, *)
class Where: Stage {
  let name: String = "where"

  let bridge: StageBridge
  private var condition: BooleanExpr

  init(condition: BooleanExpr) {
    self.condition = condition
    bridge = WhereStageBridge(expr: condition.toBridge())
  }
}

@available(iOS 13, tvOS 13, macOS 10.15, macCatalyst 13, watchOS 7, *)
class Limit: Stage {
  let name: String = "limit"

  let bridge: StageBridge
  private var limit: Int32

  init(_ limit: Int32) {
    self.limit = limit
    bridge = LimitStageBridge(limit: NSInteger(limit))
  }
}

@available(iOS 13, tvOS 13, macOS 10.15, macCatalyst 13, watchOS 7, *)
class Offset: Stage {
  let name: String = "offset"

  let bridge: StageBridge
  private var offset: Int32

  init(_ offset: Int32) {
    self.offset = offset
    bridge = OffsetStageBridge(offset: NSInteger(offset))
  }
}

@available(iOS 13, tvOS 13, macOS 10.15, macCatalyst 13, watchOS 7, *)
class AddFields: Stage {
  let name: String = "addFields"
  let bridge: StageBridge
  private var fields: [Selectable]

  init(fields: [Selectable]) {
    self.fields = fields
    let objc_accumulators = fields.reduce(into: [String: ExprBridge]()) {
      result,
        field
      in
      let seletable = field as! SelectableWrapper
      result[seletable.alias] = seletable.expr.toBridge()
    }
    bridge = AddFieldsStageBridge(fields: objc_accumulators)
  }
}

@available(iOS 13, tvOS 13, macOS 10.15, macCatalyst 13, watchOS 7, *)
class RemoveFieldsStage: Stage {
  let name: String = "removeFields"
  let bridge: StageBridge
  private var fields: [String]

  init(fields: [String]) {
    self.fields = fields
    bridge = RemoveFieldsStageBridge(fields: fields)
  }

  init(fields: [Field]) {
    self.fields = fields.map { $0.fieldName }
    bridge = RemoveFieldsStageBridge(fields: self.fields)
  }
}

@available(iOS 13, tvOS 13, macOS 10.15, macCatalyst 13, watchOS 7, *)
class Select: Stage {
  let name: String = "select"
  let bridge: StageBridge
  private var selections: [Selectable]

  init(selections: [Selectable]) {
    self.selections = selections
    let map = Helper.selectablesToMap(selectables: selections)
    bridge = SelectStageBridge(selections: map
      .mapValues { Helper.sendableToExpr($0).toBridge() })
  }
}

@available(iOS 13, tvOS 13, macOS 10.15, macCatalyst 13, watchOS 7, *)
class Distinct: Stage {
  let name: String = "distinct"
  let bridge: StageBridge
  private var groups: [Selectable]

  init(groups: [Selectable]) {
    self.groups = groups
    let map = Helper.selectablesToMap(selectables: groups)
    bridge = DistinctStageBridge(groups: map
      .mapValues { Helper.sendableToExpr($0).toBridge() })
  }
}

@available(iOS 13, tvOS 13, macOS 10.15, macCatalyst 13, watchOS 7, *)
class Aggregate: Stage {
  let name: String = "aggregate"
  let bridge: StageBridge
  private var accumulators: [AggregateWithAlias]
  private var groups: [String: Expr] = [:]

  init(accumulators: [AggregateWithAlias], groups: [Selectable]?) {
    self.accumulators = accumulators
    if groups != nil {
      self.groups = Helper.selectablesToMap(selectables: groups!)
    }
    let map = accumulators
      .reduce(into: [String: AggregateFunctionBridge]()) { result, accumulator in
        result[accumulator.alias] = accumulator.aggregate.bridge
      }
    bridge = AggregateStageBridge(
      accumulators: map,
      groups: self.groups.mapValues { Helper.sendableToExpr($0).toBridge() }
    )
  }
}

@available(iOS 13, tvOS 13, macOS 10.15, macCatalyst 13, watchOS 7, *)
class FindNearest: Stage {
  let name: String = "findNearest"
  let bridge: StageBridge
  private var field: Field
  private var vectorValue: [Double]
  private var distanceMeasure: DistanceMeasure
  private var limit: Int?
  private var distanceField: String?

  init(field: Field,
       vectorValue: [Double],
       distanceMeasure: DistanceMeasure,
       limit: Int? = nil,
       distanceField: String? = nil) {
    self.field = field
    self.vectorValue = vectorValue
    self.distanceMeasure = distanceMeasure
    self.limit = limit
    self.distanceField = distanceField
    bridge = FindNearestStageBridge(
      field: field.bridge as! FieldBridge,
      vectorValue: VectorValue(vectorValue),
      distanceMeasure: distanceMeasure.kind.rawValue,
      limit: limit as NSNumber?,
      distanceField: distanceField
    )
  }
}

@available(iOS 13, tvOS 13, macOS 10.15, macCatalyst 13, watchOS 7, *)
class Sort: Stage {
  let name: String = "sort"
  let bridge: StageBridge
  private var orderings: [Ordering]

  init(orderings: [Ordering]) {
    self.orderings = orderings
    bridge = SortStageBridge(orderings: orderings.map { $0.bridge })
  }
}

@available(iOS 13, tvOS 13, macOS 10.15, macCatalyst 13, watchOS 7, *)
class ReplaceWith: Stage {
  let name: String = "replaceWith"
  let bridge: StageBridge
  private var expr: Expr

  init(expr: Expr) {
    self.expr = expr
    bridge = ReplaceWithStageBridge(expr: expr.toBridge())
  }
}

@available(iOS 13, tvOS 13, macOS 10.15, macCatalyst 13, watchOS 7, *)
class Sample: Stage {
  let name: String = "sample"
  let bridge: StageBridge
  private var count: Int64?
  private var percentage: Double?

  init(count: Int64) {
    self.count = count
    percentage = nil
    bridge = SampleStageBridge(count: count)
  }

  init(percentage: Double) {
    self.percentage = percentage
    count = nil
    bridge = SampleStageBridge(percentage: percentage)
  }
}

@available(iOS 13, tvOS 13, macOS 10.15, macCatalyst 13, watchOS 7, *)
class Union: Stage {
  let name: String = "union"
  let bridge: StageBridge
  private var other: Pipeline

  init(other: Pipeline) {
    self.other = other
    bridge = UnionStageBridge(other: other.bridge)
  }
}

@available(iOS 13, tvOS 13, macOS 10.15, macCatalyst 13, watchOS 7, *)
class Unnest: Stage {
  let name: String = "unnest"
  let bridge: StageBridge
  private var field: Selectable
  private var indexField: String?

  init(field: Selectable, indexField: String? = nil) {
    self.field = field
    self.indexField = indexField
    bridge = UnnestStageBridge(
      field: Helper.sendableToExpr(field).toBridge(),
      indexField: indexField
    )
  }
}

@available(iOS 13, tvOS 13, macOS 10.15, macCatalyst 13, watchOS 7, *)
class RawStage: Stage {
  let name: String
  let bridge: StageBridge
  private var params: [Sendable]
  private var options: [String: Sendable]?

  init(name: String, params: [Sendable], options: [String: Sendable]? = nil) {
    self.name = name
    self.params = params
    self.options = options
    let bridgeParams = params.map { Helper.sendableToExpr($0).toBridge() }
    let bridgeOptions = options?.mapValues { Helper.sendableToExpr($0).toBridge() }
    bridge = RawStageBridge(name: name, params: bridgeParams, options: bridgeOptions)
  }
}

File: Firestore/core/src/api/stages.cc
/*
 * Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "Firestore/core/src/api/stages.h"

#include <algorithm>
#include <memory>
#include <stdexcept>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

#include "Firestore/Protos/nanopb/google/firestore/v1/document.nanopb.h"
#include "Firestore/core/src/api/pipeline.h"
#include "Firestore/core/src/core/expressions_eval.h"
#include "Firestore/core/src/model/document.h"
#include "Firestore/core/src/model/document_key.h"
#include "Firestore/core/src/model/mutable_document.h"
#include "Firestore/core/src/model/resource_path.h"
#include "Firestore/core/src/model/value_util.h"
#include "Firestore/core/src/nanopb/message.h"
#include "Firestore/core/src/nanopb/nanopb_util.h"
#include "Firestore/core/src/util/comparison.h"
#include "Firestore/core/src/util/hard_assert.h"

namespace firebase {
namespace firestore {
namespace api {

using model::DeepClone;

CollectionSource::CollectionSource(std::string path)
    : path_(model::ResourcePath::FromStringView(path)) {
}

google_firestore_v1_Pipeline_Stage CollectionSource::to_proto() const {
  google_firestore_v1_Pipeline_Stage result;

  result.name = nanopb::MakeBytesArray(name());

  result.args_count = 1;
  result.args = nanopb::MakeArray<google_firestore_v1_Value>(1);
  result.args[0].which_value_type =
      google_firestore_v1_Value_reference_value_tag;
  result.args[0].reference_value = nanopb::MakeBytesArray(
      util::StringFormat("/%s", this->path_.CanonicalString()));

  result.options_count = 0;
  result.options = nullptr;

  return result;
}

google_firestore_v1_Pipeline_Stage DatabaseSource::to_proto() const {
  google_firestore_v1_Pipeline_Stage result;

  result.name = nanopb::MakeBytesArray(name());
  result.args_count = 0;
  result.args = nullptr;
  result.options_count = 0;
  result.options = nullptr;

  return result;
}

google_firestore_v1_Pipeline_Stage CollectionGroupSource::to_proto() const {
  google_firestore_v1_Pipeline_Stage result;

  result.name = nanopb::MakeBytesArray(name());

  result.args_count = 2;
  result.args = nanopb::MakeArray<google_firestore_v1_Value>(2);
  // First argument is an empty reference value.
  result.args[0].which_value_type =
      google_firestore_v1_Value_reference_value_tag;
  result.args[0].reference_value = nanopb::MakeBytesArray("");

  // Second argument is the collection ID (encoded as a string value).
  result.args[1].which_value_type = google_firestore_v1_Value_string_value_tag;
  result.args[1].string_value = nanopb::MakeBytesArray(collection_id_);

  result.options_count = 0;
  result.options = nullptr;

  return result;
}

google_firestore_v1_Pipeline_Stage DocumentsSource::to_proto() const {
  google_firestore_v1_Pipeline_Stage result;

  result.name = nanopb::MakeBytesArray(name());

  result.args_count = static_cast<pb_size_t>(documents_.size());
  result.args = nanopb::MakeArray<google_firestore_v1_Value>(result.args_count);

  size_t i = 0;
  for (const auto& document : documents_) {
    result.args[i].which_value_type =
        google_firestore_v1_Value_reference_value_tag;
    result.args[i].reference_value = nanopb::MakeBytesArray(document);
    i++;
  }

  result.options_count = 0;
  result.options = nullptr;

  return result;
}

google_firestore_v1_Pipeline_Stage AddFields::to_proto() const {
  google_firestore_v1_Pipeline_Stage result;
  result.name = nanopb::MakeBytesArray(name());

  result.args_count = 1;
  result.args = nanopb::MakeArray<google_firestore_v1_Value>(1);

  result.args[0].which_value_type = google_firestore_v1_Value_map_value_tag;
  nanopb::SetRepeatedField(
      &result.args[0].map_value.fields, &result.args[0].map_value.fields_count,
      fields_, [](const std::pair<std::string, std::shared_ptr<Expr>>& entry) {
        return _google_firestore_v1_MapValue_FieldsEntry{
            nanopb::MakeBytesArray(entry.first), entry.second->to_proto()};
      });

  result.options_count = 0;
  result.options = nullptr;
  return result;
}

google_firestore_v1_Pipeline_Stage AggregateStage::to_proto() const {
  google_firestore_v1_Pipeline_Stage result;
  result.name = nanopb::MakeBytesArray(name());

  result.args_count = 2;
  result.args = nanopb::MakeArray<google_firestore_v1_Value>(2);

  // Encode accumulators map.
  result.args[0].which_value_type = google_firestore_v1_Value_map_value_tag;
  nanopb::SetRepeatedField(
      &result.args[0].map_value.fields, &result.args[0].map_value.fields_count,
      this->accumulators_,
      [](const std::pair<std::string, std::shared_ptr<AggregateFunction>>&
             entry) {
        return _google_firestore_v1_MapValue_FieldsEntry{
            nanopb::MakeBytesArray(entry.first), entry.second->to_proto()};
      });

  // Encode groups map.
  result.args[1].which_value_type = google_firestore_v1_Value_map_value_tag;
  nanopb::SetRepeatedField(
      &result.args[1].map_value.fields, &result.args[1].map_value.fields_count,
      this->groups_,
      [](const std::pair<std::string, std::shared_ptr<Expr>>& entry) {
        return _google_firestore_v1_MapValue_FieldsEntry{
            nanopb::MakeBytesArray(entry.first), entry.second->to_proto()};
      });

  result.options_count = 0;
  result.options = nullptr;
  return result;
}

google_firestore_v1_Pipeline_Stage Where::to_proto() const {
  google_firestore_v1_Pipeline_Stage result;

  result.name = nanopb::MakeBytesArray(name());

  result.args_count = 1;
  result.args = nanopb::MakeArray<google_firestore_v1_Value>(1);
  result.args[0] = this->expr_->to_proto();

  result.options_count = 0;
  result.options = nullptr;

  return result;
}

google_firestore_v1_Value FindNearestStage::DistanceMeasure::proto() const {
  google_firestore_v1_Value result;
  result.which_value_type = google_firestore_v1_Value_string_value_tag;
  switch (measure_) {
    case EUCLIDEAN:
      result.string_value = nanopb::MakeBytesArray("euclidean");
      break;
    case COSINE:
      result.string_value = nanopb::MakeBytesArray("cosine");
      break;
    case DOT_PRODUCT:
      result.string_value = nanopb::MakeBytesArray("dot_product");
      break;
  }
  return result;
}

google_firestore_v1_Pipeline_Stage FindNearestStage::to_proto() const {
  google_firestore_v1_Pipeline_Stage result;
  result.name = nanopb::MakeBytesArray(name());

  result.args_count = 3;
  result.args = nanopb::MakeArray<google_firestore_v1_Value>(3);
  result.args[0] = property_->to_proto();
  result.args[1] = *DeepClone(*vector_).release();
  result.args[2] = distance_measure_.proto();

  nanopb::SetRepeatedField(
      &result.options, &result.options_count, options_,
      [](const std::pair<std::string, google_firestore_v1_Value>& entry) {
        return _google_firestore_v1_Pipeline_Stage_OptionsEntry{
            nanopb::MakeBytesArray(entry.first), entry.second};
      });

  return result;
}

google_firestore_v1_Pipeline_Stage LimitStage::to_proto() const {
  google_firestore_v1_Pipeline_Stage result;
  result.name = nanopb::MakeBytesArray(name());

  result.args_count = 1;
  result.args = nanopb::MakeArray<google_firestore_v1_Value>(1);
  result.args[0].which_value_type = google_firestore_v1_Value_integer_value_tag;
  result.args[0].integer_value = limit_;

  result.options_count = 0;
  result.options = nullptr;
  return result;
}

google_firestore_v1_Pipeline_Stage OffsetStage::to_proto() const {
  google_firestore_v1_Pipeline_Stage result;
  result.name = nanopb::MakeBytesArray(name());

  result.args_count = 1;
  result.args = nanopb::MakeArray<google_firestore_v1_Value>(1);
  result.args[0].which_value_type = google_firestore_v1_Value_integer_value_tag;
  result.args[0].integer_value = offset_;

  result.options_count = 0;
  result.options = nullptr;
  return result;
}

google_firestore_v1_Pipeline_Stage SelectStage::to_proto() const {
  google_firestore_v1_Pipeline_Stage result;
  result.name = nanopb::MakeBytesArray(name());

  result.args_count = 1;
  result.args = nanopb::MakeArray<google_firestore_v1_Value>(1);

  result.args[0].which_value_type = google_firestore_v1_Value_map_value_tag;
  nanopb::SetRepeatedField(
      &result.args[0].map_value.fields, &result.args[0].map_value.fields_count,
      fields_, [](const std::pair<std::string, std::shared_ptr<Expr>>& entry) {
        return _google_firestore_v1_MapValue_FieldsEntry{
            nanopb::MakeBytesArray(entry.first), entry.second->to_proto()};
      });

  result.options_count = 0;
  result.options = nullptr;
  return result;
}

google_firestore_v1_Pipeline_Stage SortStage::to_proto() const {
  google_firestore_v1_Pipeline_Stage result;
  result.name = nanopb::MakeBytesArray(name());

  result.args_count = static_cast<pb_size_t>(orders_.size());
  result.args = nanopb::MakeArray<google_firestore_v1_Value>(result.args_count);

  for (size_t i = 0; i < orders_.size(); ++i) {
    result.args[i] = orders_[i].to_proto();
  }

  result.options_count = 0;
  result.options = nullptr;
  return result;
}

google_firestore_v1_Pipeline_Stage DistinctStage::to_proto() const {
  google_firestore_v1_Pipeline_Stage result;
  result.name = nanopb::MakeBytesArray(name());

  result.args_count = 1;
  result.args = nanopb::MakeArray<google_firestore_v1_Value>(1);

  result.args[0].which_value_type = google_firestore_v1_Value_map_value_tag;
  nanopb::SetRepeatedField(
      &result.args[0].map_value.fields, &result.args[0].map_value.fields_count,
      groups_, [](const std::pair<std::string, std::shared_ptr<Expr>>& entry) {
        return _google_firestore_v1_MapValue_FieldsEntry{
            nanopb::MakeBytesArray(entry.first), entry.second->to_proto()};
      });

  result.options_count = 0;
  result.options = nullptr;
  return result;
}

google_firestore_v1_Pipeline_Stage RemoveFieldsStage::to_proto() const {
  google_firestore_v1_Pipeline_Stage result;
  result.name = nanopb::MakeBytesArray(name());

  result.args_count = static_cast<pb_size_t>(fields_.size());
  result.args = nanopb::MakeArray<google_firestore_v1_Value>(result.args_count);

  for (size_t i = 0; i < fields_.size(); ++i) {
    result.args[i] = fields_[i].to_proto();
  }

  result.options_count = 0;
  result.options = nullptr;
  return result;
}

google_firestore_v1_Value ReplaceWith::ReplaceMode::to_proto() const {
  google_firestore_v1_Value result;
  result.which_value_type = google_firestore_v1_Value_string_value_tag;
  switch (mode_) {
    case FULL_REPLACE:
      result.string_value = nanopb::MakeBytesArray("full_replace");
      break;
    case MERGE_PREFER_NEST:
      result.string_value = nanopb::MakeBytesArray("merge_prefer_nest");
      break;
  }
  return result;
}

google_firestore_v1_Pipeline_Stage ReplaceWith::to_proto() const {
  google_firestore_v1_Pipeline_Stage result;
  result.name = nanopb::MakeBytesArray(name());

  result.args_count = 2;
  result.args = nanopb::MakeArray<google_firestore_v1_Value>(2);
  result.args[0] = expr_->to_proto();

  result.args[1] = mode_.to_proto();

  result.options_count = 0;
  result.options = nullptr;
  return result;
}

ReplaceWith::ReplaceWith(std::shared_ptr<Expr> expr, ReplaceMode mode)
    : expr_(std::move(expr)), mode_(mode) {
}

google_firestore_v1_Value Sample::SampleMode::to_proto() const {
  google_firestore_v1_Value result;
  result.which_value_type = google_firestore_v1_Value_string_value_tag;
  switch (mode_) {
    case DOCUMENTS:
      result.string_value = nanopb::MakeBytesArray("documents");
      break;
    case PERCENT:
      result.string_value = nanopb::MakeBytesArray("percent");
      break;
  }
  return result;
}

Sample::Sample(SampleMode mode, int64_t count, double percentage)
    : mode_(mode), count_(count), percentage_(percentage) {
}

google_firestore_v1_Pipeline_Stage Sample::to_proto() const {
  google_firestore_v1_Pipeline_Stage result;
  result.name = nanopb::MakeBytesArray(name());

  result.args_count = 2;
  result.args = nanopb::MakeArray<google_firestore_v1_Value>(2);

  switch (mode_.mode()) {
    case SampleMode::Mode::DOCUMENTS:
      result.args[0].which_value_type =
          google_firestore_v1_Value_integer_value_tag;
      result.args[0].integer_value = count_;
      break;
    case SampleMode::Mode::PERCENT:
      result.args[0].which_value_type =
          google_firestore_v1_Value_double_value_tag;
      result.args[0].double_value = percentage_;
      break;
  }

  result.args[1] = mode_.to_proto();

  result.options_count = 0;
  result.options = nullptr;
  return result;
}

Union::Union(std::shared_ptr<Pipeline> other) : other_(std::move(other)) {
}

google_firestore_v1_Pipeline_Stage Union::to_proto() const {
  google_firestore_v1_Pipeline_Stage result;
  result.name = nanopb::MakeBytesArray(name());

  result.args_count = 1;
  result.args = nanopb::MakeArray<google_firestore_v1_Value>(1);
  result.args[0] = other_->to_proto();

  result.options_count = 0;
  result.options = nullptr;
  return result;
}

Unnest::Unnest(std::shared_ptr<Expr> field,
               std::shared_ptr<Expr> alias,
               absl::optional<std::shared_ptr<Expr>> index_field)
    : field_(std::move(field)),
      alias_(alias),
      index_field_(std::move(index_field)) {
}

google_firestore_v1_Pipeline_Stage Unnest::to_proto() const {
  google_firestore_v1_Pipeline_Stage result;
  result.name = nanopb::MakeBytesArray(name());

  result.args_count = 2;
  result.args = nanopb::MakeArray<google_firestore_v1_Value>(2);
  result.args[0] = field_->to_proto();
  result.args[1] = alias_->to_proto();

  if (index_field_.has_value()) {
    result.options_count = 1;
    result.options =
        nanopb::MakeArray<google_firestore_v1_Pipeline_Stage_OptionsEntry>(1);
    result.options[0].key = nanopb::MakeBytesArray("index_field");
    result.options[0].value = index_field_.value()->to_proto();
  } else {
    result.options_count = 0;
    result.options = nullptr;
  }

  return result;
}

RawStage::RawStage(
    std::string name,
    std::vector<google_firestore_v1_Value> params,
    std::unordered_map<std::string, std::shared_ptr<Expr>> options)
    : name_(std::move(name)),
      params_(std::move(params)),
      options_(std::move(options)) {
}

google_firestore_v1_Pipeline_Stage RawStage::to_proto() const {
  google_firestore_v1_Pipeline_Stage result;
  result.name = nanopb::MakeBytesArray(name());

  result.args_count = static_cast<pb_size_t>(params_.size());
  result.args = nanopb::MakeArray<google_firestore_v1_Value>(result.args_count);

  for (size_t i = 0; i < result.args_count; i++) {
    result.args[i] = params_[i];
  }

  nanopb::SetRepeatedField(
      &result.options, &result.options_count, options_,
      [](const std::pair<std::string, std::shared_ptr<Expr>>& entry) {
        return _google_firestore_v1_Pipeline_Stage_OptionsEntry{
            nanopb::MakeBytesArray(entry.first), entry.second->to_proto()};
      });

  return result;
}

model::PipelineInputOutputVector CollectionSource::Evaluate(
    const EvaluateContext& /*context*/,
    const model::PipelineInputOutputVector& inputs) const {
  model::PipelineInputOutputVector results;
  std::copy_if(inputs.begin(), inputs.end(), std::back_inserter(results),
               [this](const model::MutableDocument& doc) {
                 return doc.is_found_document() &&
                        doc.key().path().PopLast().CanonicalString() ==
                            path_.CanonicalString();
               });
  return results;
}

model::PipelineInputOutputVector CollectionGroupSource::Evaluate(
    const EvaluateContext& /*context*/,
    const model::PipelineInputOutputVector& inputs) const {
  model::PipelineInputOutputVector results;
  std::copy_if(inputs.begin(), inputs.end(), std::back_inserter(results),
               [this](const model::MutableDocument& doc) {
                 return doc.is_found_document() &&
                        doc.key().GetCollectionGroup() == collection_id_;
               });
  return results;
}

model::PipelineInputOutputVector DatabaseSource::Evaluate(
    const EvaluateContext& /*context*/,
    const model::PipelineInputOutputVector& inputs) const {
  model::PipelineInputOutputVector results;
  std::copy_if(inputs.begin(), inputs.end(), std::back_inserter(results),
               [](const model::MutableDocument& doc) {
                 return doc.is_found_document();
               });
  return results;
}

model::PipelineInputOutputVector DocumentsSource::Evaluate(
    const EvaluateContext& /*context*/,
    const model::PipelineInputOutputVector& inputs) const {
  model::PipelineInputOutputVector results;
  for (const model::PipelineInputOutput& input : inputs) {
    if (input.is_found_document() &&
        documents_.count(input.key().path().CanonicalString()) > 0) {
      results.push_back(input);
    }
  }
  return results;
}

model::PipelineInputOutputVector Where::Evaluate(
    const EvaluateContext& context,
    const model::PipelineInputOutputVector& inputs) const {
  model::PipelineInputOutputVector results;
  const auto evaluable_expr = expr_->ToEvaluable();
  const auto true_value = model::TrueValue();

  for (const auto& doc : inputs) {
    auto result = evaluable_expr->Evaluate(context, doc);
    if (!result.IsErrorOrUnset() &&
        model::Equals(*result.value(), true_value)) {
      results.push_back(doc);
    }
  }

  return results;
}

model::PipelineInputOutputVector LimitStage::Evaluate(
    const EvaluateContext& /*context*/,
    const model::PipelineInputOutputVector& inputs) const {
  model::PipelineInputOutputVector::const_iterator begin;
  model::PipelineInputOutputVector::const_iterator end;
  size_t count;

  if (limit_ < 0) {
    // if limit_ is negative, we treat it as limit to last, returns the last
    // limit_ documents.
    count = static_cast<size_t>(-limit_);
    if (count > inputs.size()) {
      count = inputs.size();
    }
    begin = inputs.end() - count;
    end = inputs.end();
  } else {
    count = static_cast<size_t>(limit_);
    if (count > inputs.size()) {
      count = inputs.size();
    }
    begin = inputs.begin();
    end = inputs.begin() + count;
  }

  return model::PipelineInputOutputVector(begin, end);
}

model::PipelineInputOutputVector SortStage::Evaluate(
    const EvaluateContext& context,
    const model::PipelineInputOutputVector& inputs) const {
  model::PipelineInputOutputVector input_copy = inputs;
  std::sort(
      input_copy.begin(), input_copy.end(),
      [this, &context](const model::PipelineInputOutput& left,
                       const model::PipelineInputOutput& right) -> bool {
        for (const auto& ordering : this->orders_) {
          const auto left_result =
              ordering.expr()->ToEvaluable()->Evaluate(context, left);
          const auto right_result =
              ordering.expr()->ToEvaluable()->Evaluate(context, right);

          auto left_val = left_result.IsErrorOrUnset() ? model::MinValue()
                                                       : *left_result.value();
          auto right_val = right_result.IsErrorOrUnset()
                               ? model::MinValue()
                               : *right_result.value();
          const auto compare_result = model::Compare(left_val, right_val);
          if (compare_result != util::ComparisonResult::Same) {
            return ordering.direction() == Ordering::ASCENDING
                       ? compare_result == util::ComparisonResult::Ascending
                       : compare_result == util::ComparisonResult::Descending;
          }
        }

        return false;
      });

  return input_copy;
}

}  // namespace api
}  // namespace firestore
}  // namespace firebase
